longhaul/longhaul.md

```diff
@@ -0,0 +1,235 @@
+Fields in this structure can be reported (R), desired (D), or both.
+
+| Field Name    | type    | required | Desired or Reported | Description |
+|---------------|---------|----------|---------------------|-------------|
+| scenario      | String  | yes      | Reported            | Name of the test scenario being run 
+| totalDuration | Time    | yes      | Both                | Total run time for the test scenario being run
+
+Also, for each operation, there is a sub-object which contains configuration for that operation:
+
+| Field Name     | type    | required | Desired or Reported | Description |
+|----------------|---------|----------|---------------------|-------------|
+| enabled        | bool    | yes      | Both                | Is this op inclued in the test run?
+| interval       | integer | yes      | Both                | interval (in seconds) to run this test operation
+| opsPerInterval | integer | yes      | Both                | How many test operations to run per interval
+
+## Fields (progress)
+
+The data in the progress object describes the progress of teh test run along with a minimal set of 
+telemetry that can be used to graph the progression of the test
+
+| Field Name            | type    | required | Description |
+|-----------------------|---------|----------|-------------|
+| index                 | integer | yes      | index for this progress record, starts at 1 and incriments for each succcessive report
+| status                | string  | yes      | current status.  one of new, running, failed, or succeeded 
+| processUptime         | Time    | yes      | Amount of time the process has been running
+| memoryUsed            | double  | yes      | Amount of heap use in the process in MB
+| activeObjects         | integer | no       | Number of active objects being used (if available)
+| completeOperations    | integer | yes      | Number of discrete opersations run so far
+| outstandingOperations | integer | yes      | Number of operations started but not yet completed
+| slowOperations        | integer | yes      | Number of operations that completed "slowly" (defined by the slowness threshold for that operation type)
+
+## Fields (stats)
+
+The data in the stats object describes performance statistics on different groups of operations.
+The stats object is composed of a number of sub objects, each named for the test operation they apply to.
+
+| Field Name        | type    | required | Description |
+|-------------------|---------|----------|-------------|
+| totalComplete     | integer | yes      | Total number of operations of this type completed so far
+| totalOutstanding  | integer | yes      | Total number of operations of this type started but not completed
+| slownessThreshold | doule   | yes      | Minimum duration, in seconds, that causes an operation to be considered slow
+| totalSlow         | integer | yes      | Total number of operations that are considered "slow"
+
```

Comment:
is there a way to report the error type(s) that are seen as stats as well?

Reply:
I'm curious to hear how you think we should categorize errors for reporting.  At the very least, we can discriminate between API failures and timeouts (e.g. doesn't complete in 5 minutes).  Maybe we can have this be free form, the name of the Exception object, which would differ between Node and Python.  

---
In reply to: [442351670](https://github.com/Azure/azure-iot-sdk-python/pull/595#discussion_r442351670) [](ancestors = 442351670)